### PR TITLE
Fix service restarts when puppet agent is enabled

### LINKED ISSUE #32: Service restarts always when puppet agent is enabled
When puppet agent is enabled, the service restarts always when the puppet agent polling happens

### PR DESCRIPTION
**Pull request by vidurananayakkara**
Resolves #32

### CODE REVIEW

File: hieradata/dev/wso2/common.yaml [hunk 1]
```diff
@@ -55,6 +55,9 @@ wso2::template_list:
 #wso2::file_list:
 #  - repository/conf/security/cipher-tool.properties
 #  - repository/conf/security/cipher-text.properties
+#  - repository/conf/security/secret-conf.properties
+
+#wso2::file_list_copy_without_refresh:
```
Comment (imesh) [inline hieradata/dev/wso2/common.yaml]: It might be better to rename this variable name. Shall we call it "service_refresh_file_list" and add a comment explaining its behaviour?

File: manifests/clean.pp [hunk 2]
```diff
@@ -15,42 +15,6 @@
...
-    'zero': {
-      exec { "stop_process_remove_CARBON_HOME_and_pack_${carbon_home}":
-        path    => '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/java/bin/',
-        command => "kill -9 `cat ${carbon_home}/wso2carbon.pid` && rm -rf ${carbon_home} && rm -f ${pack_dir}/${pack_filename}";
-      }
-    }
-
-    default: { fail("Clean mode ${mode} is not supported by this module") }
-  }
+  # TODO: clean is not required and is to be removed from the upcoming versions of puppet modules
```
Comment (imesh) [inline manifests/clean.pp]: May be we can change the text "from the upcoming versions" to "from the next major version release". WDYT?

File: manifests/service.pp [hunk 3]
```diff
@@ -21,6 +21,17 @@
...
+      $file_list_copy_without_refresh:
+        owner       => $wso2_user,
+        group       => $wso2_group,
+        carbon_home => $carbon_home,
+        wso2_module => $caller_module_name
+    }
+  }
 
   # Start the service
   # TODO: start the service only if configuration changes are applied that needs a restart to be effective
```
Comment (imesh) [inline manifests/service.pp]: Since this PR implements this feature this TODO statement can be removed IMO.

**Review by imesh — CHANGES_REQUESTED** [review]

File: manifests/service.pp [hunk 4]
```diff
@@ -21,6 +21,17 @@
...
   $autostart_service = $wso2base::autostart_service
+  $file_list_copy_without_refresh = $wso2base::file_list_copy_without_refresh
+
+  if ($file_list_copy_without_refresh != undef and size($file_list_copy_without_refresh) > 0) {
+    wso2base::push_files {
+      $file_list_copy_without_refresh:
+        owner       => $wso2_user,
+        group       => $wso2_group,
+        carbon_home => $carbon_home,
+        wso2_module => $caller_module_name
```
Comment (chamilad) [inline manifests/service.pp]: Doesn't this need a `require` field or an addition of a `require` field to some other resource to make sure Puppet picks up the correct order in which this action should be executed?

**Review by chamilad — CHANGES_REQUESTED** [review]
Verify the need of a `require` field

File: manifests/service.pp (hunk 4 above)
Comment (vidurananayakkara) [reply manifests/service.pp]: According to [1] by default puppet applies resources in the order they are declared in the manifest. The execution order would be wso2base::push_files -> Service[$service_name] -> notify. According to my understanding, this is not required.

[1] https://docs.puppet.com/puppet/latest/lang_relationships.html

Comment (chamilad) [reply manifests/service.pp]: I'm more concerned with the effect of keeping this without binding to any other resource that would result in this resource being executed parallelly to some other resource. It would be better to be explicit in the required order.

File: hieradata/dev/wso2/common.yaml [hunk 5]
```diff
@@ -51,10 +51,14 @@ wso2::template_list:
...
 #  - bin/ciphertool.sh
 
-# Puppet file list to be copied
+# Puppet file list to be copied triggering restart
 #wso2::file_list:
 #  - repository/conf/security/cipher-tool.properties
 #  - repository/conf/security/cipher-text.properties
+#  - repository/conf/security/secret-conf.properties
+
+# Puppet file list to be copied without restarting the service
```
Comment (imesh) [inline hieradata/dev/wso2/common.yaml]: Isn't this giving a different meaning? Wouldn't it be more appropriate to say "Files to be copied when restarting the system service"? In some other location we might need to mention that the system service is only restarted if there are distribution changes or configuration changes available.

File: hieradata/dev/wso2/common.yaml [hunk 6]
```diff
@@ -51,10 +51,14 @@ wso2::template_list:
   - bin/wso2server.sh
 #  - bin/ciphertool.sh
 
-# Puppet file list to be copied
+# Puppet file list to be copied triggering restart
```
Comment (imesh) [inline hieradata/dev/wso2/common.yaml]: I think these are not puppet files. They can be any files used in the product. 

Moreover, it might be better to make this sentence more clear, shall we rephrase this to something similar to "File list to be copied to the product folder, if there are changes found in the files, the system service will be restarted"?

File: manifests/clean.pp [hunk 7]
```diff
@@ -14,43 +14,7 @@
 #  limitations under the License.
 #----------------------------------------------------------------------------
 
+# TODO: Clean will be removed from the next major version release of puppet modules
```
Comment (imesh) [inline manifests/clean.pp]: Shouldn't it be "wso2base puppet module" instead of "puppet modules"?

**Review by chamilad — APPROVED** [review]
LGTM

**Review by imesh — APPROVED** [review]
